Type issue with TypeScript 4.9

<!-- If this is more of a question on how i18n should work or you have not narrowed down the issue to be 1 of our packages, please use Discussions instead -->

## Which package?

`@formatjs/intl`

## Describe the bug

These type issues started appearing after I upgraded to TypeScript 4.9.

```
node_modules/@formatjs/intl/src/types.d.ts:21:26 - error TS2339: Property 'ids' does not exist on type 'Message'.

21 } ? FormatjsIntl.Message['ids'] : string;
                            ~~~~~

node_modules/@formatjs/intl/src/types.d.ts:24:29 - error TS2339: Property 'locale' does not exist on type 'IntlConfig'.

24 } ? FormatjsIntl.IntlConfig['locale'] : string;
```

## To Reproduce

### Codesandbox URL

Please include a codesandbox URL if possible

### Reproducible Steps/Repo

Steps to reproduce the behavior:

1. Install [TypeScript 4.9.3](https://devblogs.microsoft.com/typescript/announcing-typescript-4-9/)
2. run `tsc`

## Expected behavior

No type issues.

## Screenshots
If applicable, add screenshots to help explain your problem.

## Desktop (please complete the following information):

- OS: [e.g. iOS] Windows 11
- Browser [e.g. chrome, safari] n/a
- Version [e.g. 22] `@formatjs/intl@2.5.1`

## Smartphone (please complete the following information):

- Device: [e.g. iPhone6] n/a
- OS: [e.g. iOS8.1] n/a
- Browser [e.g. stock browser, safari] n/a
- Version [e.g. 22] n/a

## Additional context

I did not have this issue with TS 4.8.

--- a/package.json
+++ b/package.json
@@ -114,6 +114,7 @@
     "ts-jest": "28",
     "ts-loader": "^9.1.2",
     "ts-node": "^10.8",
+    "tsd": "^0.24.1",
     "tslib": "2.4.0",
     "typescript": "^4.7",
     "unidiff": "^1.0.2",

--- a/packages/intl/BUILD
+++ b/packages/intl/BUILD
@@ -1,8 +1,10 @@
 load("@aspect_bazel_lib//lib:write_source_files.bzl", "write_source_files")
 load("@aspect_rules_js//npm/private:npm_package.bzl", "npm_package")
+load("@npm//:tsd/package_json.bzl", tsd_bin = "bin")
 load("@npm//:defs.bzl", "npm_link_all_packages")
 load("//tools:index.bzl", "check_format", "package_json_test", "ts_compile")
 load("//tools:jest.bzl", "jest_test")
+load("@aspect_bazel_lib//lib:copy_to_directory.bzl", "copy_to_directory")
 
 npm_link_all_packages(name = "node_modules")
 
@@ -76,4 +78,32 @@ check_format(
 package_json_test(
     name = "package_json_test",
     deps = SRC_DEPS,
+)
+
+# TODO: make this a macro that paramtrizes on `.test-d.ts` files.
+copy_to_directory(
+    name = "global_type_overrides_test_files",
+    srcs = [
+        ":%s" % PACKAGE_NAME,
+        "tests/global_type_overrides.test-d.ts"
+    ],
+    out = "type_test_files",
+    # Remove the prefix so the package directory merges with the test file directory.
+    replace_prefixes = dict([
+        ("%s" % PACKAGE_NAME, "")
+    ])
+)
+
+tsd_bin.tsd_test(
+    name = "global_type_overrides_test",
+    data = [
+        "//:node_modules/tsd",
+        ":global_type_overrides_test_files",
+        "tsconfig.json",
+    ] + SRC_DEPS,
+    args = [
+        "--files",
+        "tests/**/*.test-d.ts"
+    ],
+    chdir = "packages/%s/type_test_files" % PACKAGE_NAME
 )

--- a/packages/intl/src/types.ts
+++ b/packages/intl/src/types.ts
@@ -23,6 +23,9 @@ import {
 import {DEFAULT_INTL_CONFIG} from './utils'
 import {NumberFormatOptions} from '@formatjs/ecma402-abstract'
 
+// Note: FormatjsIntl is defined as a global namespace so the library user can
+// override the default types of Message.ids (e.g. as string literal unions from extracted strings)
+// or IntlConfig.locale (e.g. to a list of supported locales).
 declare global {
   namespace FormatjsIntl {
     interface Message {}
@@ -31,12 +34,14 @@ declare global {
   }
 }
 
-type MessageIds = FormatjsIntl.Message extends {ids: string}
-  ? FormatjsIntl.Message['ids']
-  : string
+// NOTE: workaround the TypeScript 4.9 bug: https://github.com/formatjs/formatjs/issues/3910
+type _Message = FormatjsIntl.Message
+type MessageIds = _Message extends {ids: string} ? _Message['ids'] : string
 
-type Locale = FormatjsIntl.IntlConfig extends {locale: string}
-  ? FormatjsIntl.IntlConfig['locale']
+// NOTE: workaround the TypeScript 4.9 bug: https://github.com/formatjs/formatjs/issues/3910
+type _IntlConfig = FormatjsIntl.IntlConfig
+type Locale = _IntlConfig extends {locale: string}
+  ? _IntlConfig['locale']
   : string
 
 export type OnErrorFn = (

--- a/packages/intl/tests/global_type_overrides.test-d.ts
+++ b/packages/intl/tests/global_type_overrides.test-d.ts
@@ -1,0 +1,18 @@
+import {expectType} from 'tsd'
+import {MessageDescriptor, ResolvedIntlConfig} from '../src/types'
+
+// Example type overrides
+declare global {
+  namespace FormatjsIntl {
+    interface Message {
+      ids: 'a' | 'b'
+    }
+    interface IntlConfig {
+      locale: 'en-US' | 'zh-CN'
+    }
+  }
+}
+
+// Check that the type overrides actually work.
+expectType<'a' | 'b'>(null as any as NonNullable<MessageDescriptor['id']>)
+expectType<'en-US' | 'zh-CN'>(null as any as ResolvedIntlConfig['locale'])

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -27,6 +27,9 @@
     "jsx": "react"
   },
   "exclude": [
-    "packages/react-intl/example-sandboxes"
+    "packages/react-intl/example-sandboxes",
+    // Exclude any tsd tests because they can mess with normal
+    // type checking in editors.
+    "**/*.test-d.ts"
   ]
 }

